Conditional boundary event trigger tests

berkaycanbc opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

Closes #43537

berkaycanbc committed:
feat: integrate conditional evaluation on variable changes

berkaycanbc committed:
feat: integrate conditional evaluation on variable changes

berkaycanbc committed:
test: conditional boundary event trigger on variable changes

berkaycanbc committed:
refactor: implement review changes

berkaycanbc committed:
refactor: implement review changes

berkaycanbc committed:
refactor: optimize expression parsing

lzgabel reviewed (changes requested):
❤️  @berkaycanbc. Thank you very much for adding detailed code comments. They helped me better understand the intent of the test cases during the review.

Left some comments, please take a look. 🙇

lzgabel commented:
@berkaycanbc Merry Christmas in advance—hope you have a joyful holiday and lots of fun! 🎅🎁

berkaycanbc commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/conditional/triggering/ConditionalBoundaryEventTriggerTest.java:
```diff
@@ -0,0 +1,1882 @@
+            .serviceTask(SIBLING_TASK_ID, t -> t.zeebeJobType("sibling"))
+            .parallelGateway("join")
+            .endEvent("end")
+            .moveToNode(SIBLING_TASK_ID)
```
@lzgabel We actually do that only once 😅. I've copied the output XML to modeler and result:

<img width="613" height="357" alt="Screenshot 2026-01-07 at 13 23 26" src="https://github.com/user-attachments/assets/306cc096-d2c1-48fa-a061-b94cae14f056" />

berkaycanbc replied:
@lzgabel I've actually played around with sequence flows previously but didn't see still. Now, I am seeing it 😅. Updated.

berkaycanbc commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/conditional/triggering/ConditionalBoundaryEventTriggerTest.java:
```diff
@@ -0,0 +1,1882 @@
+        .update();
+
+    // then
+    assertThat(
```
Updated 👍

berkaycanbc commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/conditional/triggering/ConditionalBoundaryEventTriggerTest.java:
```diff
@@ -0,0 +1,1882 @@
+    engine.variables().ofScope(taskInstanceKey).withDocument(Map.of("x", 11)).update();
+    engine.variables().ofScope(taskInstanceKey).withDocument(Map.of("x", 12)).update();
+
+    // then (non-interrupting boundary may trigger again)
```
> 🤔 I'm wondering if we need to add an assertion verification to ensure that the triggered record is generated after the Activity is activated, avoiding unexpected occurrences due to upgrades, migrations, or other factors.

@lzgabel This sounds like more of a topic for integration tests. Still, I did not consider to add such test cases due to the test setup overhead. But maybe we should consider them?

berkaycanbc commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/conditional/triggering/ConditionalBoundaryEventTriggerTest.java:
```diff
@@ -0,0 +1,1882 @@
+    // then - boundary triggers only once (for that child)
+    engine.signal().withSignalName("testSpeedUp").broadcast();
+
+    assertThat(
```
Good point, updated 👍

berkaycanbc committed:
refactor: improve assertions and test naming

lzgabel reviewed (approved):
👍 Thanks @berkaycanbc . 🚀  LGTM.

georgios-goulos commented:
I see there is some core logic added, so the PR title is a bit misleading.

georgios-goulos commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnConditionalBehavior.java:
nitpick: `conditionSubscriptionState` -> `conditionalSubscriptionState`

georgios-goulos commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/conditional/triggering/ConditionalBoundaryEventTriggerTest.java:
```diff
@@ -0,0 +1,1989 @@
+
+    final long taskInstanceKey = awaitElementInstance(processInstanceKey, TASK_ID);
+
+    // when (change variable z; since there is no filter, evaluation must run and trigger)
```
> change variable z

What is variable `z`?

berkaycanbc committed:
refactor: improve assertions and comments

berkaycanbc committed:
test: add multiple variable tests

berkaycanbc commented:
> I may have missed them, but are there any tests using conditions with multiple variables ("x + y > 10")? It would be good to have such tests for cases like updating only one of the vars.

@georgios-goulos I did not add them originally because it felt more like testing the expression processor. But still better than not having it. I've added a test case for the one you mentioned. Additionally, I've added simple multiple variable test cases to the trigger on activation test classes.

georgios-goulos reviewed (approved):
Good tests 👍
